File: combomaker.js
// List of martial arts moves
const moves = [
  'jab',
  'jab (body)',
  'hook',
  'hook (body)',
  'lead uppercut',
  'lead elbow',
  'slip left',
  'reset',
  'switch kick',
  'teep',
  'switch knee',
  'cross',
  'cross (body)',
  'rear hook / overhand',
  'rear hook (body)',
  'rear uppercut',
  'rear elbow',
  'slip right',
  'rear kick',
  'rear teep',
  'rear knee'
];

// each move, and the set of moves that can follow it
const rulesDict = {
  'jab': {
    'jab': true, 'jab (body)': true, 'hook': false, 'hook (body)': false, 'lead uppercut': false, 'lead elbow': false, 'slip left': true, 'reset': true, 'switch kick': false, 'teep': false, 'switch knee': false, 'cross': true, 'cross (body)': true, 'rear hook / overhand': true, 'rear hook (body)': true, 'rear uppercut': true, 'rear elbow': true, 'slip right': false, 'rear kick': true, 'rear teep': true, 'rear knee': true
  },
  'jab (body)': {
    'jab': true, 'jab (body)': true, 'hook': false, 'hook (body)': false, 'lead uppercut': false, 'lead elbow': false, 'slip left': true, 'reset': true, 'switch kick': false, 'teep': false, 'switch knee': false, 'cross': true, 'cross (body)': true, 'rear hook / overhand': true, 'rear hook (body)': true, 'rear uppercut': true, 'rear elbow': true, 'slip right': false, 'rear kick': true, 'rear teep': true, 'rear knee': true
  },
  'hook': {
    'jab': false, 'jab (body)': false, 'hook': false, 'hook (body)': false, 'lead uppercut': false, 'lead elbow': false, 'slip left': true, 'reset': true, 'switch kick': false, 'teep': false, 'switch knee': false, 'cross': true, 'cross (body)': true, 'rear hook / overhand': true, 'rear hook (body)': true, 'rear uppercut': true, 'rear elbow': true, 'slip right': false, 'rear kick': true, 'rear teep': true, 'rear knee': true
  },
  'hook (body)': {
    'jab': false, 'jab (body)': false, 'hook': false, 'hook (body)': false, 'lead uppercut': false, 'lead elbow': false, 'slip left': true, 'reset': true, 'switch kick': false, 'teep': false, 'switch knee': false, 'cross': true, 'cross (body)': true, 'rear hook / overhand': true, 'rear hook (body)': true, 'rear uppercut': true, 'rear elbow': true, 'slip right': false, 'rear kick': true, 'rear teep': true, 'rear knee': true
  },
  'lead uppercut': {
    'jab': false, 'jab (body)': false, 'hook': false, 'hook (body)': false, 'lead uppercut': false, 'lead elbow': false, 'slip left': true, 'reset': true, 'switch kick': false, 'teep': false, 'switch knee': false, 'cross': true, 'cross (body)': true, 'rear hook / overhand': true, 'rear hook (body)': true, 'rear uppercut': true, 'rear elbow': true, 'slip right': false, 'rear kick': true, 'rear teep': true, 'rear knee': true
  },
  'lead elbow': {
    'jab': false, 'jab (body)': false, 'hook': false, 'hook (body)': false, 'lead uppercut': false, 'lead elbow': false, 'slip left': true, 'reset': true, 'switch kick': false, 'teep': false, 'switch knee': false, 'cross': true, 'cross (body)': true, 'rear hook / overhand': true, 'rear hook (body)': true, 'rear uppercut': true, 'rear elbow': true, 'slip right': false, 'rear kick': true, 'rear teep': true, 'rear knee': true
  },
  'slip left': {
    'jab': true, 'jab (body)': true, 'hook': true, 'hook (body)': true, 'lead uppercut': true, 'lead elbow': true, 'slip left': false, 'reset': false, 'switch kick': true, 'teep': false, 'switch knee': true, 'cross': false, 'cross (body)': false, 'rear hook / overhand': false, 'rear hook (body)': false, 'rear uppercut': false, 'rear elbow': false, 'slip right': false, 'rear kick': false, 'rear teep': false, 'rear knee': false
  },
  'reset': {
    'jab': true, 'jab (body)': true, 'hook': true, 'hook (body)': true, 'lead uppercut': true, 'lead elbow': true, 'slip left': false, 'reset': false, 'switch kick': true, 'teep': true, 'switch knee': true, 'cross': true, 'cross (body)': true, 'rear hook / overhand': true, 'rear hook (body)': true, 'rear uppercut': true, 'rear elbow': true, 'slip right': false, 'rear kick': true, 'rear teep': true, 'rear knee': true
  },
  'switch kick': {
    'jab': false, 'jab (body)': false, 'hook': false, 'hook (body)': false, 'lead uppercut': false, 'lead elbow': false, 'slip left': true, 'reset': true, 'switch kick': false, 'teep': false, 'switch knee': false, 'cross': true, 'cross (body)': true, 'rear hook / overhand': true, 'rear hook (body)': false, 'rear uppercut': true, 'rear elbow': false, 'slip right': true, 'rear kick': false, 'rear teep': false, 'rear knee': true
  },
  'teep': {
    'jab': false, 'jab (body)': false, 'hook': false, 'hook (body)': false, 'lead uppercut': false, 'lead elbow': false, 'slip left': true, 'reset': true, 'switch kick': false, 'teep': false, 'switch knee': false, 'cross': true, 'cross (body)': true, 'rear hook / overhand': true, 'rear hook (body)': true, 'rear uppercut': true, 'rear elbow': false, 'slip right': false, 'rear kick': false, 'rear teep': false, 'rear knee': false
  },
  'switch knee': {
    'jab': false, 'jab (body)': false, 'hook': false, 'hook (body)': false, 'lead uppercut': false, 'lead elbow': false, 'slip left': true, 'reset': true, 'switch kick': false, 'teep': false, 'switch knee': false, 'cross': true, 'cross (body)': true, 'rear hook / overhand': true, 'rear hook (body)': false, 'rear uppercut': true, 'rear elbow': false, 'slip right': true, 'rear kick': false, 'rear teep': false, 'rear knee': true
  },
  'cross': {
    'jab': false, 'jab (body)': false, 'hook': true, 'hook (body)': true, 'lead uppercut': true, 'lead elbow': true, 'slip left': false, 'reset': true, 'switch kick': true, 'teep': true, 'switch knee': true, 'cross': false, 'cross (body)': false, 'rear hook / overhand': false, 'rear hook (body)': false, 'rear uppercut': false, 'rear elbow': false, 'slip right': true, 'rear kick': false, 'rear teep': false, 'rear knee': false
  },
  'cross (body)': {
    'jab': false, 'jab (body)': false, 'hook': true, 'hook (body)': true, 'lead uppercut': true, 'lead elbow': true, 'slip left': false, 'reset': true, 'switch kick': true, 'teep': true, 'switch knee': true, 'cross': false, 'cross (body)': false, 'rear hook / overhand': false, 'rear hook (body)': false, 'rear uppercut': false, 'rear elbow': false, 'slip right': true, 'rear kick': false, 'rear teep': false, 'rear knee': false
  },
  'rear hook / overhand': {
    'jab': false, 'jab (body)': false, 'hook': true, 'hook (body)': true, 'lead uppercut': true, 'lead elbow': true, 'slip left': false, 'reset': true, 'switch kick': true, 'teep': true, 'switch knee': true, 'cross': false, 'cross (body)': false, 'rear hook / overhand': false, 'rear hook (body)': false, 'rear uppercut': false, 'rear elbow': false, 'slip right': true, 'rear kick': false, 'rear teep': false, 'rear knee': false
  },
  'rear hook (body)': {
    'jab': true, 'jab (body)': true, 'hook': true, 'hook (body)': true, 'lead uppercut': true, 'lead elbow': true, 'slip left': false, 'reset': true, 'switch kick': true, 'teep': true, 'switch knee': true, 'cross': false, 'cross (body)': false, 'rear hook / overhand': false, 'rear hook (body)': false, 'rear uppercut': false, 'rear elbow': false, 'slip right': true, 'rear kick': false, 'rear teep': false, 'rear knee': false
  },
  'rear uppercut': {
    'jab': false, 'jab (body)': false, 'hook': true, 'hook (body)': true, 'lead uppercut': true, 'lead elbow': true, 'slip left': false, 'reset': true, 'switch kick': true, 'teep': true, 'switch knee': true, 'cross': false, 'cross (body)': false, 'rear hook / overhand': false, 'rear hook (body)': false, 'rear uppercut': false, 'rear elbow': false, 'slip right': true, 'rear kick': false, 'rear teep': false, 'rear knee': false
  },
  'rear elbow': {
    'jab': false, 'jab (body)': false, 'hook': true, 'hook (body)': true, 'lead uppercut': true, 'lead elbow': true, 'slip left': false, 'reset': true, 'switch kick': true, 'teep': true, 'switch knee': true, 'cross': false, 'cross (body)': false, 'rear hook / overhand': false, 'rear hook (body)': false, 'rear uppercut': false, 'rear elbow': false, 'slip right': true, 'rear kick': false, 'rear teep': false, 'rear knee': false
  },
  'slip right': {
    'jab': false, 'jab (body)': false, 'hook': false, 'hook (body)': false, 'lead uppercut': false, 'lead elbow': false, 'slip left': true, 'reset': true, 'switch kick': false, 'teep': false, 'switch knee': false, 'cross': true, 'cross (body)': true, 'rear hook / overhand': true, 'rear hook (body)': true, 'rear uppercut': true, 'rear elbow': true, 'slip right': false, 'rear kick': true, 'rear teep': true, 'rear knee': true
  },
  'rear kick': {
    'jab': false, 'jab (body)': false, 'hook': true, 'hook (body)': true, 'lead uppercut': true, 'lead elbow': true, 'slip left': false, 'reset': true, 'switch kick': true, 'teep': true, 'switch knee': true, 'cross': false, 'cross (body)': false, 'rear hook / overhand': false, 'rear hook (body)': false, 'rear uppercut': false, 'rear elbow': false, 'slip right': true, 'rear kick': false, 'rear teep': false, 'rear knee': false
  },
  'rear teep': {
    'jab': true, 'jab (body)': false, 'hook': true, 'hook (body)': true, 'lead uppercut': false, 'lead elbow': false, 'slip left': false, 'reset': false, 'switch kick': false, 'teep': false, 'switch knee': false, 'cross': false, 'cross (body)': false, 'rear hook / overhand': true, 'rear hook (body)': true, 'rear uppercut': false, 'rear elbow': true, 'slip right': false, 'rear kick': true, 'rear teep': true, 'rear knee': true
  },
  'rear knee': {
    'jab': true, 'jab (body)': true, 'hook': true, 'hook (body)': true, 'lead uppercut': false, 'lead elbow': true, 'slip left': false, 'reset': true, 'switch kick': false, 'teep': true, 'switch knee': false, 'cross': false, 'cross (body)': false, 'rear hook / overhand': false, 'rear hook (body)': false, 'rear uppercut': false, 'rear elbow': false, 'slip right': true, 'rear kick': false, 'rear teep': false, 'rear knee': false
  }
};

const generateCombination = (moves, rulesDict, comboLength) => {
  const combination = [];
  let prevMove = null;
  while (combination.length < comboLength) {
    const move = moves[Math.floor(Math.random() * moves.length)];
    if (prevMove === null) {
      combination.push(move);
      prevMove = move;
    } else if (rulesDict[prevMove][move]) {
      combination.push(move);
      prevMove = move;
    }
  }
  return combination;
};

const getCombo = () => {
    const comboLength = document.getElementById("combo-length").value;
    const includeReset = document.getElementById("include-reset").checked;
    const includeSlips = document.getElementById("include-slips").checked;

    // Filter moves based on toggles
    let filteredMoves = moves.filter(move => {
        if (!includeReset && move === "reset") return false;
        if (!includeSlips && (move === "slip left" || move === "slip right")) return false;
        return true;
    });

    // Generate combination using filtered moves
    const combination = generateCombination(filteredMoves, rulesDict, comboLength);

    // Update the result in the HTML
    var resultDiv = document.getElementById("result");
    resultDiv.innerHTML = "";  // Clear previous result

    // Append each move to the result
    for (var i = 0; i < combination.length; i++) {
        var move = ""+(i+1) +". "+ combination[i];
        var moveElement = document.createElement("p");
        moveElement.innerText = move;
        resultDiv.appendChild(moveElement);
    }
}

File: combomaker.py
import random

comboLength = length = int(input("Length of the combination: ") or 5)

def generate_combination(moves, rules):
    combination = []
    prev_move = None
    while len(combination) < comboLength:
        move = random.choice(moves)
        if prev_move is None:
            combination.append(move)
            prev_move = move
        if prev_move is not None and rules[moves.index(prev_move)][moves.index(move)]:
            combination.append(move)
            prev_move = move

    
    return combination

# List of martial arts moves
moves = [
    'jab',
    'body jab',
    'hook',
    'body hook',
    'lead uppercut',
    'lead elbow',
    'slipL',
    'resetL',
    'switch kick',
    'teep',
    'switch knee',
    'cross',
    'body cross',
    'overhand',
    'rear body hook',
    'rear uppercut',
    'rear elbow',
    'slipR',
    'resetR',
    'rear kick',
    'rear teep',
    'rear knee',
    'roll'
]

# Move rules as a list of lists
# List of martial arts moves
moves = [
    'jab', 'body jab', 'hook', 'body hook', 'lead uppercut', 'lead elbow', 'slipL',
    'reset', 'switch kick', 'teep', 'switch knee', 'cross', 'body cross', 'overhand',
    'rear body hook', 'rear uppercut', 'rear elbow', 'slipR', 'rear kick',
    'rear teep', 'rear knee'
]

# Move rules as a list of lists. Each row represents a move, the booleans in each combo say which moves _can follow_.
rules = [
    # jab   bod-jab hook    body-hk  lead-up  lead-el  slipL   reset   sw-k   teep   sw-kn  cross  bodCrs  overh  r-bod-hk r-up  rr-el slipR  r-k    r-teep r-kn 
    [True , True ,  False,  False,   False,   False,   True ,  True ,  False, False, False, True,  True,   True,  True,    True, True, False, True,  True,  True],
    [True , True ,  False,  False,   False,   False,   True ,  True ,  False, False, False, True,  True,   True,  True,    True, True, False, True,  True,  True],
    [False, False,  False,  False,   False,   False,   True ,  True ,  False, False, False, True,  True,   True,  True,    True, True, False, True,  True,  True],
    [False, False,  False,  False,   False,   False,   True ,  True ,  False, False, False, True,  True,   True,  True,    True, True, False, True,  True,  True],
    [False, False,  False,  False,   False,   False,   True ,  True ,  False, False, False, True,  True,   True,  True,    True, True, False, True,  True,  True],
    [False, False,  False,  False,   False,   False,   True ,  True ,  False, False, False, True,  True,   True,  True,    True, True, False, True,  True,  True],
    # slip left
    [True , True ,  True,   True,    True,    True,    False , False,  True,  False, True , False, False,  False, False,   False,False,False, False, False, False],
    # reset
    [True , True ,  True,   True,    True,    True,    False , False,  True,  True , True , True , True ,  True , True ,   True ,True ,False, True , True , True ],
    # switch kick
    [False, False,  False,  False,   False,   False,   True ,  True ,  False, False, False, True,  True,   True,  True,    False,True, False, False, True,  False],
    # teep
    [False, False,  False,  False,   False,   False,   True ,  True ,  False, False, False, True,  True,   True,  True,    True, True, False, False, False, False],
    [False, False,  False,  False,   False,   False,   True ,  True ,  False, False, False, True,  True,   True,  True,    False,True, False, False, True,  False],
    # cross
    [False, False,  True,   True,    True,    True,    False , True,   True,  True,  True , False, False,  False, False,   False,False,True,  False, False, False],
    [False, False,  True,   True,    True,    True,    False , True,   True,  True,  True , False, False,  False, False,   False,False,True,  False, False, False],
    # overhand
    [False, False,  True,   True,    True,    True,    False , True,   True,  True,  True , False, False,  False, False,   False,False,True,  False, False, False],
    [False, False,  True,   True,    True,    True,    False , True,   True,  True,  True , False, False,  False, False,   False,False,True,  False, False, False],
    [False, False,  True,   True,    True,    True,    False , True,   True,  True,  True , False, False,  False, False,   False,False,True,  False, False, False],
    [False, False,  True,   True,    True,    True,    False , True,   True,  True,  True , False, False,  False, False,   False,False,True,  False, False, False],
    # slip right
    [False, False,  False,  False,   False,   False,   False,  False,  False, False,  False, True, True,   True,  True,    True, True, False, True,  False, True],
    # rear kick
    [True,  True,   True,   True,    False,   True,    False,  True ,  False, True,   False, False, False, False, False,   False,False,True,  False, False, False],
    # rear teep
    [True,  True,   True,   True,    True,    True,    False,  True ,  False, False,  False, False, False, False, False,   False,False,True,  False, False, False],
    [True,  True,   True,   True,    False,   True,    False,  True ,  False, True ,  False, False, False, False, False,   False,False,True,  False, False, False]
]

# Generate a combination of moves
combination = generate_combination(moves, rules)

# Print the combination
for move in combination:
    print(move)